extensions/markdown/src/extension.ts

```diff
@@ -265,4 +266,33 @@ class MDDocumentContentProvider implements TextDocumentContentProvider {
 			}, 300);
 		}
 	}
+}
+
+class DocumentHeadingsProvider implements vscode.DocumentSymbolProvider {
+
+	// http://daringfireball.net/projects/markdown/syntax#header
+	private static _atxPattern = /^(#){1,6}\s+.+(\s+\1)?/;
```

Comment:
On the regex:
* Strip out any number of trailing `#`s at the end of the line. This does not have to match the number at the start: http://spec.commonmark.org/0.26/#example-42
* Match end of line too, to handle weird cases like `# a # a #`
* Strip out whitespace around the header text itself for the displayed symbol.

Something like: `^(#{1,6})\s+(.+?)\s*#*\s*$` and use these two captures to build up the symbol